src/machine/machine.c

```diff
@@ -61,21 +59,29 @@ int machine_new(Manager *manager, MachineClass class, const char *name, Machine
         if (!m->name)
                 return -ENOMEM;
 
-        if (class != MACHINE_HOST) {
-                m->state_file = path_join("/run/systemd/machines", m->name);
-                if (!m->state_file)
+        m->class = class;
+
+        *ret = TAKE_PTR(m);
+        return 0;
+}
+
+int machine_link(Manager *manager, Machine *machine) {
+        int r;
+        assert(manager);
+        assert(machine);
+
+        if (machine->class != MACHINE_HOST) {
+                machine->state_file = path_join("/run/systemd/machines", machine->name);
+                if (!machine->state_file)
                         return -ENOMEM;
         }
 
-        m->class = class;
-
-        r = hashmap_put(manager->machines, m->name, m);
+        r = hashmap_put(manager->machines, machine->name, machine);
         if (r < 0)
                 return r;
 
-        m->manager = manager;
+        machine->manager = manager;
 
-        *ret = TAKE_PTR(m);
         return 0;
 }
```

Comment:
hmm, so if call machine_link on a Machine object and it fails for some reason, and i call it a 2nd time later, it might leak the `state_file` field.

of course, this is never actually going to happy, but we generally like functions that either succeed in full, or not at all, so that you can call them multiple times like this without ill effect. – at least when it is easy to do.

hence, maybe add a check up top that refuses things early if `machine->manager` is already set up. And maybe free `state_file` if it's already set (i.e. use free_and_strdup() or so)